packages/jamtools/features/modules/dashboards/keytar_and_foot_dashboard/multi_octave_supervisor.tsx

```diff
@@ -0,0 +1,161 @@
+import React from 'react';
+
+import {ModuleAPI} from '~/core/engine/module_api';
+import {MidiEvent, MidiEventFull} from '~/core/modules/macro_module/macro_module_types';
+import {Button} from '~/core/components/Button';
+
+type MultiOctaveSupervisorMidiState = {
+    currentlyHeldDownInputNotes: MidiEvent[];
+};
+
+export class MultiOctaveSupervisor {
+    private macros!: Awaited<ReturnType<MultiOctaveSupervisor['createMacros']>>;
+    private states!: Awaited<ReturnType<MultiOctaveSupervisor['createStates']>>;
+    private actions!: ReturnType<MultiOctaveSupervisor['createActions']>;
+
+    private midiState: MultiOctaveSupervisorMidiState = {
+        currentlyHeldDownInputNotes: [],
+    };
+
+    private handleKeyboardNote = async (fullEvent: MidiEventFull) => {
+        const event = fullEvent.event;
+
+        this.macros.pagedOctaveOutput.send(event);
+
+        if (event.type === 'noteon') {
+            if (!this.midiState.currentlyHeldDownInputNotes.find(e => e.number === event.number)) {
+                this.midiState = {
+                    ...this.midiState,
+                    currentlyHeldDownInputNotes: [
+                        ...this.midiState.currentlyHeldDownInputNotes,
+                        event,
+                    ],
+                };
+            }
+        }
+
+        if (event.type === 'noteoff') {
+            this.midiState = {
+                ...this.midiState,
+                currentlyHeldDownInputNotes: this.midiState.currentlyHeldDownInputNotes.filter(e => {
+                    return e.number !== event.number;
+                }),
+            };
+        }
+
+        setTimeout(() => {
+            if (this.states.enableDebugging.getState()) {
+                this.states.debugSavedInputEvent.setState(fullEvent);
+                this.states.debugMidiState.setState(this.midiState);
+            }
+        });
+    };
+
+    private createActions = () => ({
+        toggleDebugging: this.moduleAPI.createAction(`${this.kvPrefix}|toggleDebugging`, {}, async () => {
+            console.log('toggling debug mode', !this.states.enableDebugging.getState());
+            this.states.enableDebugging.setState(!this.states.enableDebugging.getState());
+        }),
+    });
```

Comment:
These actions magically run on only the Maestro device. They are essentially HTTP requests sent to the Maestro device. In the case where the Maestro device _is_ the browser that did the UI thing (e.g. the user is playing in local mode or hybrid mode), it just runs the function immediately when the action is called.

The action system mainly exists so you can do stuff in the React UI and trigger code to run on the Maestro device. Any state changes you want to show on multiple devices, that were caused by doing something in the UI, should usually be done in an action.

Behind the scenes it's using websockets and RPC, but the developer doesn't need to know that, similar to how we trust that React renders stuff that we give it. It's a mystery unless you read their source code.

The `jenkem` framework creates the RPC action behind the scenes, when the function `moduleAPI.createAction` is called on line 55 above. It does all the RPC stuff based on the context of the module API and some other things. More info in the docs `jenkem` framework docs [here](https://docs.jam.tools/typedoc_docs/module_api/classes/ModuleAPI/#createaction).

References for JSON RPC info:
https://en.wikipedia.org/wiki/Remote_procedure_call
https://en.wikipedia.org/wiki/JSON-RPC#Examples
https://www.jsonrpc.org/specification